### PR TITLE
Add -S option to pvget/pvmonitor to stringify byte arrays

### PR DESCRIPTION
**Pull request by JJL772**
This emulates the behavior of the caget -S option. Byte arrays are displayed as strings.

Example (without -S):
```
$ pvget SIOC:TST:SYS0:APP_DIR
SIOC:TST:SYS0:APP_DIR 2026-03-17 23:09:56.242  [47,109,101,100,105,97,47,66,105,103,68,114,105,118,101,47,80,114,111,106,101,99,116,115,47,101,112,105,99,115,47,105,111,99,47,97,116,108,97,115,45,103,105,116,47,105,111,99,66,111,111,116,47,115,105,111,99,45,116,115,116,45,115,121,115,48,0]
```

Example (with -S):
```
$ pvget -S SIOC:TST:SYS0:APP_DIR
SIOC:TST:SYS0:APP_DIR 2026-03-17 22:39:05.669  /media/BigDrive/Projects/epics/ioc/atlas-git/iocBoot/sioc-tst-sys0
```

pvAccess PR: https://github.com/epics-base/pvAccessCPP/pull/222

### CODE REVIEW

**Comment by anjohnson** [comment]
Appveyor builds with older VMs need a newer version of pip. Nothing to do with this PR, I think we need to fix the CI in Base.

**Comment by ralphlange** [comment]
Nothing to do with this PR - agreed.

The CI builds of Base don't have this issue - it's the CI configuration of _this_ module that needs fixing.

**Comment by JJL772** [comment]
Just realized I forgot to open the corresponding PR in pvAccess to actually add the `-S` option...: https://github.com/epics-base/pvAccessCPP/pull/222

File: src/pv/pvData.h [hunk 1]
```diff
@@ -906,6 +908,8 @@ class epicsShareClass PVStructure : public PVField, public BitSetSerializable
 
         FORCE_INLINE Formatter& format(format_t f) { xfmt = f; return *this; }
 
+        FORCE_INLINE Formatter& asString(bool b) { xstring = b; return *this; }
```
Comment (mdavidsaver) [inline src/pv/pvData.h]: `asString()` seems too generic for this quite specific "treat byte array as string"  modifier.

Comment (JJL772) [reply src/pv/pvData.h]: ive renamed it to `arrayAsString` to be a bit more specific

File: src/factory/printer.cpp [hunk 2]
```diff
@@ -434,12 +434,36 @@ std::ostream& operator<<(std::ostream& strm, const PVStructure::Formatter& forma
...
+                    if (format.xarrAsStr && scalarArray && scalarArray->getElementType() == pvByte) {
+                        shared_vector<const int8_t> byteArray;
+                        pvArray->getAs(byteArray);
+
+                        const char* str = reinterpret_cast<const char*>(byteArray.data());
+                        const size_t slen = strlen(str);
+
+                        const size_t buflen = epicsStrnEscapedFromRawSize(str, slen);
+
+                        char* buf = new char[buflen+1] {};
```
Comment (mdavidsaver) [inline src/factory/printer.cpp]: RAII please!

As simple as `std::vector<char> buf(buflen+1);`, then `&buf[0]` to reference.

Comment (JJL772) [reply src/factory/printer.cpp]: Done. I also fixed an error where an empty (?) scalar array would cause a crash, as `str` would be NULL.

File: src/factory/printer.cpp [hunk 3]
```diff
@@ -434,12 +434,37 @@ std::ostream& operator<<(std::ostream& strm, const PVStructure::Formatter& forma
...
+                    ScalarArrayConstPtr scalarArray = pvArray->getScalarArray();
+
+                    // Stringify byte arrays if requested
+                    if (format.xarrAsStr && scalarArray && scalarArray->getElementType() == pvByte) {
+                        shared_vector<const int8_t> byteArray;
+                        pvArray->getAs(byteArray);
+
+                        const char* str = reinterpret_cast<const char*>(byteArray.data());
+                        if (str) {
+                            const size_t slen = strlen(str);
```
Comment (mdavidsaver) [inline src/factory/printer.cpp]: With `shared_vector`, a better pattern is:

```c++
if(!byteArray.empty()) {
    const char* str = reinterpret_cast<const char*>(byteArray.data());
    const size_t slen = byteArray.size();
```

A `strlen()` is largely redundant.  And the slight redundancy would hide any nulls if this byteArray is actually just a byte array, and not intended as a string.  I think it would be legit to hide the last byte if it is null, which would be the case for a printable string, but not null bytes in the middle.

Comment (JJL772) [reply src/factory/printer.cpp]: my intent was to stop at the first NULL terminator, similar to how `caget` behaves:
```
$ caput -a mystring 5 33 0 33 33 33
$ caget -S mystring
mystring !
```
Without the `strlen`, strings with more than one trailing byte would also end up getting escaped (though I don't know how often this sort of thing would happen in practice).

Comment (mdavidsaver) [reply src/factory/printer.cpp]: > ... though I don't know how often this sort of thing would happen in practice).

User error?  If the PV being inspected is not actually a long string, or if the server gets the size wrong.  Both situations I would want to see.  To my mind the CLI tools have an important role in this sort of troubleshooting, before anyone is aware that there is a problem.

Also, I realize that you have not been testing the edge case where there are characters to escape ;)  (me neither of course, but that is less fun to point out!)  Any added escapes need more bytes.  This is where `epicsStrnEscapedFromRawSize()` comes in.  You can find a couple of usages in Base.  eg `caget.c`

Putting it all together, I think what I am suggesting would look like:

```c++
if(!byteArray.empty()) {
    const char* str = reinterpret_cast<const char*>(byteArray.data());
    const size_t slen = byteArray.size();
    if(slen && str[slen-1]=='\0')
        slen--; // ignore one optional trailing null
    std::vector<char> buf(epicsStrnEscapedFromRawSize(str, slen)+1);
    epicsStrnEscapedFromRaw(buf.data(), buf.size(), str, slen);
```

Comment (mdavidsaver) [reply src/factory/printer.cpp]: Ha, I am forgetting that instead of `epicsStrnEscapedFromRaw()`, someone added an iostream aware `epics::pvData::escape()`.  (I wonder who that was...)

```c++
if(!byteArray.empty()) {
    const char* str = reinterpret_cast<const char*>(byteArray.data());
    const size_t slen = byteArray.size();
    if(slen && str[slen-1]=='\0')
        slen--; // ignore one optional trailing null
    strm<<escape(std::string(str, slen));
```

https://github.com/epics-base/pvDataCPP/blob/13b56e704c1c6e3428cc6345057e16be02f9ac2f/testApp/misc/testprinter.cpp#L338-L345

Comment (JJL772) [reply src/factory/printer.cpp]: I have applied the suggested changes
